awscli/customizations/arguments.py

```diff
@@ -51,3 +54,66 @@ def override_required_args(self, argument_table, args, **kwargs):
         if name_in_cmdline in args:
             for arg_name in argument_table.keys():
                 argument_table[arg_name].required = False
+
+
+class StatefulArgument(CustomArgument):
+    """An argument that maintains a stateful value"""
+
+    def __init__(self, *args, **kwargs):
+        super(StatefulArgument, self).__init__(*args, **kwargs)
+        self._value = None
+
+    def add_to_params(self, parameters, value):
+        self._value = value
+
+    @property
+    def value(self):
+        return self._value
+
+
+class OutFileArgument(StatefulArgument):
+    """An argument that persists a result to a file"""
+
+    def add_to_params(self, parameters, value):
+        # Validate the file here so we can raise an error prior
+        # calling the service.
+        if value is None:
+            # Only validate the path if it was provided.
+            return
+        outfile = os.path.expandvars(value)
+        outfile = os.path.expanduser(outfile)
+        # Ensure that the file is writable before executing a command.
+        if not os.access(os.path.dirname(os.path.abspath(outfile)), os.W_OK):
+            raise ValueError('Unable to write to file: %s' % outfile)
+        self._value = outfile
+
+
+class QueryOutFileArgument(OutFileArgument):
+    """An argument that write a JMESPath query result to a file"""
+
+    def __init__(self, name, query, *args, **kwargs):
+        self._query = query
+        # Generate default help_text if text was not provided.
+        if 'help_text' not in kwargs:
+            kwargs['help_text'] = ('Saves the command output contents of %s '
+                                   'to the given filename' % self.query)
+        super(QueryOutFileArgument, self).__init__(name, *args, **kwargs)
+
+    @property
+    def query(self):
+        return self._query
+
+    def save_query(self, parsed_result):
+        """Saves the result of a JMESPath expression to a file.
+
+        If the argument value was not set, then the file is not saved
+        or modified.
+        """
+        if self.value and self.query:
+            contents = jmespath.search(self.query, parsed_result)
+            with open(self.value, 'wb') as fp:
+                # Don't write 'None' to a file -- write ''.
+                if contents is None:
+                    fp.write(''.encode())
+                else:
+                    fp.write(contents.encode())
```

Comment:
Given that we're writing a file to their filesystem, seems like it's more reasonable to use whatever encoding the user may have configured rather than assume utf-8. We can just open the file via `r` and and let python handle the encoding.